Adding product test for aborted transaction in Hive ACID

harmandeeps opened the pull request.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -0,0 +1,101 @@
+    // This simulates a aborted transaction which leaves behind a file in a partition of a table as follows:
+    // 1. Open Txn
+    // 2. Rollback Txn
+    // 3. Create a new table with location as the given partition location + delta directory for the rolled back transaction
```
Does it do what we want?
You are allocating write id for table A. And then you create table B for which the writeId for the given transaction was not allocated at all.
There is no relation between table B and writeId allocated so it does not seem to simulate situation that transaction for table B was aborted, and files were left behind.

harmandeeps replied:
Table A is a transactional table. It insertes some data say X, Y. And then I created a transaction and then aborted it which creates a delta.

Then, I am creating an external table B (non transactional) on the aborted transaction delta and then inserting some data (say Z) getting stored on this location. **This step is to simulate that the aborted transaction has left some data at the location.** 

Then, in the main test, I am querying Table A(transactional table), and checking that the aborted transaction delta is not getting read.

losipiuk replied:
Cool - I get it now.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -0,0 +1,101 @@
+    private static ThriftHiveMetastoreClient createMetastoreClient()
+            throws TException
+    {
+        URI metastore = URI.create("thrift://hadoop-master:9083");
```
You should not hardcode hostname and port.
Inject values into test code via:
```
    @Inject(optional = true)
    @Named("databases.hive.metastore.host")
    private String metastoreHost;

    @Inject(optional = true)
    @Named("databases.hive.metastore.port")
    private int metastorePort;
```

harmandeeps replied:
I tried, but it is not able to fetch the values defined in yml file. can you point out what am i missing?

losipiuk replied:
https://github.com/prestosql/presto/commit/0303aa63f94cc72f568f70738d195ed3e78e8385
works fine

harmandeeps replied:
thanks for the help, incorporated the changes.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -0,0 +1,101 @@
+            throws TException
+    {
+        URI metastore = URI.create("thrift://hadoop-master:9083");
+        return new ThriftHiveMetastoreClient(
```
This is a bit problematic because the way you construct client will not work for all tests setups (e.g. if metastore requires authentication).
What I think would be best to avoid explicit communication with metastore at all. Though I am not sure if we can reliably simulate failed transaction just using JDBC connection to Hive. 

....

I thought that maybe we have a thrift-client which works with authentication pre-setup in tempto. But it looks like we do not. So maybe we should just run this test in setup without authentication on hive metastore. 


@findepi thoughts?

findepi replied:
> I thought that maybe we have a thrift-client which works with authentication pre-setup in tempto. But it looks like we do not. So maybe we should just run this test in setup without authentication on hive metastore.

@losipiuk  makes sense

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -0,0 +1,101 @@
+            hiveQueryExecutor.executeQuery(String.format("CREATE EXTERNAL TABLE %s (col string) stored as ORC location '%s'", tmpTableName, deltaLocation));
+
+            // Step 4
+            hiveQueryExecutor.executeQuery(String.format("INSERT INTO TABLE %s SELECT 'a'", tmpTableName));
```
I do not like the assumption here that the layout of files generated by INSERT matches what Hive expects for a delta directory.
To validate approach  can we create another delta directory but for non-aborted transaction. And assert that rows from that directory are visible on query?

harmandeeps replied:
If I understood correctly, I have made changes to insert delta after this step to validate.

losipiuk replied:
This is not exactly my point. What I would like to have, is to add data for non-rolled-backed transaction using same approach you use for rolled-back transaction. 
With creating temporary external table pointing to delta directory, and doing insert to temporary table. 

I want to have it to validate that way we use to insert data works as expected.
Without the validation there is a chance that we are not seeing rows inserted to delta-directory for rolled-back transaction, not because the transaction is rolled back, but because test is broken :)

Is it more clear?

harmandeeps replied:
I like the approach, but creating an external table on a valid delta location, and then inserting data in ACID format with keeping the file name in format bucket_000XYZ, is not possible in product tests IMHO. Please let me know if you know a way around this.

To validate the 2nd point, I deleted the data present at a valid delta location. I created an external table on a valid delta location, inserted the data form this external table. But, then Presto select query failed with error: 

```
o.prestosql.spi.PrestoException: ORC ACID file should have 6 columns: hdfs://hadoop-master:9000/user/hive/warehouse/dummy_id_data/delta_0000001_0000001_0000/000000_0

at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.verifyAcidSchema(OrcPageSourceFactory.java:423)
```

So, this is validating our assumption that rolled back transaction won't be read from Presto, if it is being read, then the query will fail with the error mentioned above.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -355,6 +357,64 @@ public void testCreateAcidTable(boolean isPartitioned, BucketingType bucketingTy
+            String hivePartition = isPartitioned ? " PARTITION (part_col=2) " : "";
+            String predicate = isPartitioned ? " WHERE part_col = 2 " : "";
+
+            onHive().executeQuery("INSERT OVERWRITE TABLE " + tableName + hivePartition + " select 1");
+
+            String selectFromOnePartitionsSql = "SELECT col FROM " + tableName + predicate + " ORDER BY COL";
+            QueryResult onePartitionQueryResult = query(selectFromOnePartitionsSql);
+            assertThat(onePartitionQueryResult).containsOnly(row(1));
+
+            onHive().executeQuery("INSERT INTO TABLE " + tableName + hivePartition + " select 2");
+            onePartitionQueryResult = query(selectFromOnePartitionsSql);
+            assertThat(onePartitionQueryResult).containsExactly(row(1), row(2));
+
+            onHive().executeQuery("INSERT OVERWRITE TABLE " + tableName + hivePartition + " select 3");
+            onePartitionQueryResult = query(selectFromOnePartitionsSql);
+            assertThat(onePartitionQueryResult).containsOnly(row(3));
+
+            onHive().executeQuery("INSERT INTO TABLE " + tableName + hivePartition + " select 4");
+            onePartitionQueryResult = query(selectFromOnePartitionsSql);
+            assertThat(onePartitionQueryResult).containsExactly(row(3), row(4));
```
Do we need that many inserts before actually simulating aborted transaction.
It is not obvious what cases are we testing here.

harmandeeps replied:
yeah, have removed some of them.

losipiuk replied:
thanks

harmandeeps reviewed (commented):
@losipiuk : Thanks for the comments, I have addressed them, please take a look.

harmandeeps commented on presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftMetastoreClient.java:
```diff
@@ -179,4 +187,10 @@ String get_config_value(String name, String defaultValue)
     String getDelegationToken(String userName)
             throws TException;
+
+    default List<TxnToWriteId> allocateTableWriteIdsBatchIntr(AllocateTableWriteIdsRequest rqst)
```
done.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -40,40 +38,32 @@ private ACIDTestHelper()
-            assertThat(hiveQueryExecutor.executeQuery(String.format("SELECT * FROM %s", tmpTableName))).containsOnly(row("a"));
-            hiveQueryExecutor.executeQuery(String.format("DROP TABLE %s", tmpTableName));
+            // Insert data to the external table
+            QueryExecutors.onHive().executeQuery(String.format("INSERT INTO TABLE %s SELECT 'a'", tmpTableName));
```
static import `onHive`

losipiuk commented:
> I like the approach, but creating an external table on a valid delta location, and then inserting data in ACID format with keeping the file name in format bucket_000XYZ, is not possible in product tests IMHO. Please let me know if you know a way around this.
> 
> To validate the 2nd point, I deleted the data present at a valid delta location. I created an external table on a valid delta location, inserted the data form this external table. But, then Presto select query failed with error: 
> 
> ```
> o.prestosql.spi.PrestoException: ORC ACID file should have 6 columns: hdfs://hadoop-master:9000/user/hive/warehouse/dummy_id_data/delta_0000001_0000001_0000/000000_0
> 
> at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.verifyAcidSchema(OrcPageSourceFactory.java:423)
> ```
> 
> So, this is validating our assumption that rolled back transaction won't be read from Presto, if it is being read, then the query will fail with the error mentioned above.

Yeah very good point  - I did not thing about the fact that external insert format would not match the ACID table requirements. 

I would still prefer to have it tested continously not, one off manually.

The approach which could work could be:

- crate ACID table t
- insert delta A into t
- create delta directory B for a failed transaction
- create delta directory C for non failed transaction
- copy contents of A to B and C
- select from t and verify that we get the rows we inserted twice (presumably from A and C)

Actually it feels that it would be simpler than what you are doing right now. 
And also more resemble the real world, because now we essentially put garbage into delta directory for aborted transaction.

It should be easily implementable using `io.prestosql.tempto.hadoop.hdfs.HdfsClient` we have already.
You have methods for listing and creating directories, reading and writing files. Should be enough.

Would you be fine giving it a try?

findepi commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/ACIDTestHelper.java:
```diff
@@ -0,0 +1,97 @@
+import static io.prestosql.tempto.assertions.QueryAssert.Row.row;
+import static io.prestosql.tempto.assertions.QueryAssert.assertThat;
+
+public final class ACIDTestHelper
```
ACID -> Acid

harmandeeps replied:
fixed. changed it to Transactional as we have used Transactional everywhere.

harmandeeps commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -358,6 +364,42 @@ public void testCreateAcidTable(boolean isPartitioned, BucketingType bucketingTy
+    @Test(groups = HIVE_TRANSACTIONAL)
+    public void testFilesForAbortedTransactionsIgnored()
+            throws TException, IOException
+    {
```
@losipiuk : removed partition test case, I guess as we are only testing at the delta directory level, so it won't make any difference. thoughts?

harmandeeps commented:
> > I like the approach, but creating an external table on a valid delta location, and then inserting data in ACID format with keeping the file name in format bucket_000XYZ, is not possible in product tests IMHO. Please let me know if you know a way around this.
> > To validate the 2nd point, I deleted the data present at a valid delta location. I created an external table on a valid delta location, inserted the data form this external table. But, then Presto select query failed with error:
> > ```
> > o.prestosql.spi.PrestoException: ORC ACID file should have 6 columns: hdfs://hadoop-master:9000/user/hive/warehouse/dummy_id_data/delta_0000001_0000001_0000/000000_0
> > 
> > at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.verifyAcidSchema(OrcPageSourceFactory.java:423)
> > ```
> > 
> > 
> > So, this is validating our assumption that rolled back transaction won't be read from Presto, if it is being read, then the query will fail with the error mentioned above.
> 
> Yeah very good point - I did not thing about the fact that external insert format would not match the ACID table requirements.
> 
> I would still prefer to have it tested continously not, one off manually.
> 
> The approach which could work could be:
> 
> * crate ACID table t
> * insert delta A into t
> * create delta directory B for a failed transaction
> * create delta directory C for non failed transaction
> * copy contents of A to B and C
> * select from t and verify that we get the rows we inserted twice (presumably from A and C)
> 
> Actually it feels that it would be simpler than what you are doing right now.
> And also more resemble the real world, because now we essentially put garbage into delta directory for aborted transaction.
> 
> It should be easily implementable using `io.prestosql.tempto.hadoop.hdfs.HdfsClient` we have already.
> You have methods for listing and creating directories, reading and writing files. Should be enough.
> 
> Would you be fine giving it a try?

Thanks for the suggestion @losipiuk . I have incorporated the approach you mentioned. Though, simply copying the data didn't work because of ORC getting malformed while copying. So, have to create ORC reader/writer to copy. Please have a look.

losipiuk commented:
> > > I like the approach, but creating an external table on a valid delta location, and then inserting data in ACID format with keeping the file name in format bucket_000XYZ, is not possible in product tests IMHO. Please let me know if you know a way around this.
> > > To validate the 2nd point, I deleted the data present at a valid delta location. I created an external table on a valid delta location, inserted the data form this external table. But, then Presto select query failed with error:
> > > ```
> > > o.prestosql.spi.PrestoException: ORC ACID file should have 6 columns: hdfs://hadoop-master:9000/user/hive/warehouse/dummy_id_data/delta_0000001_0000001_0000/000000_0
> > > 
> > > at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.verifyAcidSchema(OrcPageSourceFactory.java:423)
> > > ```
> > > 
> > > 
> > > So, this is validating our assumption that rolled back transaction won't be read from Presto, if it is being read, then the query will fail with the error mentioned above.
> > 
> > 
> > Yeah very good point - I did not thing about the fact that external insert format would not match the ACID table requirements.
> > I would still prefer to have it tested continously not, one off manually.
> > The approach which could work could be:
> > 
> > * crate ACID table t
> > * insert delta A into t
> > * create delta directory B for a failed transaction
> > * create delta directory C for non failed transaction
> > * copy contents of A to B and C
> > * select from t and verify that we get the rows we inserted twice (presumably from A and C)
> > 
> > Actually it feels that it would be simpler than what you are doing right now.
> > And also more resemble the real world, because now we essentially put garbage into delta directory for aborted transaction.
> > It should be easily implementable using `io.prestosql.tempto.hadoop.hdfs.HdfsClient` we have already.
> > You have methods for listing and creating directories, reading and writing files. Should be enough.
> > Would you be fine giving it a try?
> 
> Thanks for the suggestion @losipiuk . I have incorporated the approach you mentioned. Though, simply copying the data didn't work because of ORC getting malformed while copying. So, have to create ORC reader/writer to copy. Please have a look.

Thanks. I like that better. Some more suggestions:
* Now logic the test and `simulateAbortedHiveTransaction` are so dependent on each other that IMO having those two separated makes code less readable. `simulateAbortedHiveTransaction` is not a generic helper method any more. I suggest that you inline `simulateAbortedHiveTransaction` into test method. Then you can have better naming for variables representing delta directories. I think it will make code read better.
* I think using ORCReader/Writer is not great as it is production code - but I can live with that given fact that we are testing another aspect of code. Why simple copying did not work? What error did you observe?

harmandeeps commented:
> > > > I like the approach, but creating an external table on a valid delta location, and then inserting data in ACID format with keeping the file name in format bucket_000XYZ, is not possible in product tests IMHO. Please let me know if you know a way around this.
> > > > To validate the 2nd point, I deleted the data present at a valid delta location. I created an external table on a valid delta location, inserted the data form this external table. But, then Presto select query failed with error:
> > > > ```
> > > > o.prestosql.spi.PrestoException: ORC ACID file should have 6 columns: hdfs://hadoop-master:9000/user/hive/warehouse/dummy_id_data/delta_0000001_0000001_0000/000000_0
> > > > 
> > > > at io.prestosql.plugin.hive.orc.OrcPageSourceFactory.verifyAcidSchema(OrcPageSourceFactory.java:423)
> > > > ```
> > > > 
> > > > 
> > > > So, this is validating our assumption that rolled back transaction won't be read from Presto, if it is being read, then the query will fail with the error mentioned above.
> > > 
> > > 
> > > Yeah very good point - I did not thing about the fact that external insert format would not match the ACID table requirements.
> > > I would still prefer to have it tested continously not, one off manually.
> > > The approach which could work could be:
> > > 
> > > * crate ACID table t
> > > * insert delta A into t
> > > * create delta directory B for a failed transaction
> > > * create delta directory C for non failed transaction
> > > * copy contents of A to B and C
> > > * select from t and verify that we get the rows we inserted twice (presumably from A and C)
> > > 
> > > Actually it feels that it would be simpler than what you are doing right now.
> > > And also more resemble the real world, because now we essentially put garbage into delta directory for aborted transaction.
> > > It should be easily implementable using `io.prestosql.tempto.hadoop.hdfs.HdfsClient` we have already.
> > > You have methods for listing and creating directories, reading and writing files. Should be enough.
> > > Would you be fine giving it a try?
> > 
> > 
> > Thanks for the suggestion @losipiuk . I have incorporated the approach you mentioned. Though, simply copying the data didn't work because of ORC getting malformed while copying. So, have to create ORC reader/writer to copy. Please have a look.
> 
> Thanks. I like that better. Some more suggestions:
> 
> * Now logic the test and `simulateAbortedHiveTransaction` are so dependent on each other that IMO having those two separated makes code less readable. `simulateAbortedHiveTransaction` is not a generic helper method any more. I suggest that you inline `simulateAbortedHiveTransaction` into test method. Then you can have better naming for variables representing delta directories. I think it will make code read better.
> * I think using ORCReader/Writer is not great as it is production code - but I can live with that given fact that we are testing another aspect of code. Why simple copying did not work? What error did you observe?

Thanks for the inputs @losipiuk . I will change to merge `simulateAbortedHiveTransaction` into the test itself.

Regarding using ORC writer, previously, I used `hdfsClient.loadFile` and then `hdfsClient.saveFile` to write to the file. But, then I got errors like `malformed orc file invalid postscript`, I tried different ways, but it didn't help. So, at last, I zeroed upon ORC writer/reader.

losipiuk commented:
> Regarding using ORC writer, previously, I used `hdfsClient.loadFile` and then `hdfsClient.saveFile` to write to the file. But, then I got errors like `malformed orc file invalid postscript`, I tried different ways, but it didn't help. So, at last, I zeroed upon ORC writer/reader.

Interesting, seems like bug in `HdfsClient`. To double check, you used the `loadFile`,`saveFile` versions which operate on byte streams (`public void loadFile(String path, OutputStream outputStream)`, `public void saveFile(String path, InputStream input)`). Not helper versions which take, return data as `String`, right?

harmandeeps commented:
> > Regarding using ORC writer, previously, I used `hdfsClient.loadFile` and then `hdfsClient.saveFile` to write to the file. But, then I got errors like `malformed orc file invalid postscript`, I tried different ways, but it didn't help. So, at last, I zeroed upon ORC writer/reader.
> 
> Interesting, seems like bug in `HdfsClient`. To double check, you used the `loadFile`,`saveFile` versions which operate on byte streams (`public void loadFile(String path, OutputStream outputStream)`, `public void saveFile(String path, InputStream input)`). Not helper versions which take, return data as `String`, right?

I used ` loadFile(String path, OutputStream outputStream)` to load file in an OutputStream, but then was not able to convert in InputStream for method `saveFile(String path, InputStream input)`, so had to use `hdfsClient.saveFile(b, new String(outputStream.toByteArray(), Charset.defaultCharset()));` and other variants. If, you know the solution I can try that.

losipiuk commented:
> > > Regarding using ORC writer, previously, I used `hdfsClient.loadFile` and then `hdfsClient.saveFile` to write to the file. But, then I got errors like `malformed orc file invalid postscript`, I tried different ways, but it didn't help. So, at last, I zeroed upon ORC writer/reader.
> > 
> > 
> > Interesting, seems like bug in `HdfsClient`. To double check, you used the `loadFile`,`saveFile` versions which operate on byte streams (`public void loadFile(String path, OutputStream outputStream)`, `public void saveFile(String path, InputStream input)`). Not helper versions which take, return data as `String`, right?
> 
> I used ` loadFile(String path, OutputStream outputStream)` to load file in an OutputStream, but then was not able to convert in InputStream for method `saveFile(String path, InputStream input)`, so had to use `hdfsClient.saveFile(b, new String(outputStream.toByteArray(), Charset.defaultCharset()));` and other variants. If, you know the solution I can try that.

Going through `String` would not work as ORC is binary format.
Something like this should work:
```java
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        client.loadFile(filePath, bos);
        client.saveFile(newPath, new ByteArrayInputStream(bos.toByteArray()));
```

harmandeeps commented:
> > > > Regarding using ORC writer, previously, I used `hdfsClient.loadFile` and then `hdfsClient.saveFile` to write to the file. But, then I got errors like `malformed orc file invalid postscript`, I tried different ways, but it didn't help. So, at last, I zeroed upon ORC writer/reader.
> > > 
> > > 
> > > Interesting, seems like bug in `HdfsClient`. To double check, you used the `loadFile`,`saveFile` versions which operate on byte streams (`public void loadFile(String path, OutputStream outputStream)`, `public void saveFile(String path, InputStream input)`). Not helper versions which take, return data as `String`, right?
> > 
> > 
> > I used ` loadFile(String path, OutputStream outputStream)` to load file in an OutputStream, but then was not able to convert in InputStream for method `saveFile(String path, InputStream input)`, so had to use `hdfsClient.saveFile(b, new String(outputStream.toByteArray(), Charset.defaultCharset()));` and other variants. If, you know the solution I can try that.
> 
> Going through `String` would not work as ORC is binary format.
> Something like this should work:
> 
> ```java
>         ByteArrayOutputStream bos = new ByteArrayOutputStream();
>         client.loadFile(filePath, bos);
>         client.saveFile(newPath, new ByteArrayInputStream(bos.toByteArray()));
> ```

Yeah, I missed this workaround somehow. Updated the PR, thanks.

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TransactionalTestHelper.java:
```diff
@@ -0,0 +1,52 @@
+import java.net.URI;
+import java.util.Optional;
+
+public final class TransactionalTestHelper
```
As it is used in just a single place now please inline into `TestHiveTransactionalTable.java`

harmandeeps replied:
It is getting used in https://github.com/prestosql/presto/pull/5404/files tests as well, so I thought of keeping the utils separately, to keep the test file shorter. thoughts?

losipiuk replied:
Then please rename the file to `TestHiveMetastoreClientFactory`

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -358,6 +376,67 @@ public void testCreateAcidTable(boolean isPartitioned, BucketingType bucketingTy
+            // Insert data to create a valid delta, which creates `delta-B`
+            onHive().executeQuery("INSERT INTO TABLE " + tableName + " SELECT 3");
+
+            // Simulate aborted transaction in Hive which has left behind a write directory and file
+            long transaction = client.openTransaction("test");
+
+            client.allocateTableWriteIds("default", tableName, Collections.singletonList(transaction)).get(0).getWriteId();
+
+            // Rollback transaction which leaves behind a delta directory `delta-C` i.e. `delta_0000003_0000003_0000`
+            client.abortTransaction(transaction);
```
```suggestion
            // Simulate aborted transaction in Hive which has left behind a write directory and file (`delta-C` i.e `delta_0000003_0000003_0000`)
            long transaction = client.openTransaction("test");
            client.allocateTableWriteIds("default", tableName, Collections.singletonList(transaction)).get(0).getWriteId();
            client.abortTransaction(transaction);
```

losipiuk commented on presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java:
```diff
@@ -358,6 +376,67 @@ public void testCreateAcidTable(boolean isPartitioned, BucketingType bucketingTy
+
+            // Copy content of `delta-A` to `delta-B`
+            hdfsClient.saveFile(deltaB, new ByteArrayInputStream(byteArrayOutputStream.toByteArray()));
+
```
```suggestion
// Verify that data from delta-A and delta-B is visible
assertThat(onePartitionQueryResult).containsOnly(row(1), row(1), row(2), row(2));
```

harmandeeps replied:
fixed.

losipiuk replied:
I meant to add extra step with just delta-A and delta-B. The comment by the last assertion was fine. Sorry for not being clear.

harmandeeps replied:
got it, fixed.

losipiuk reviewed (commented):
Thanks.

Just minor editorial remaining and we are good to go :)

losipiuk reviewed (commented):
Clarified comments. Please squash commits.

harmandeeps commented:
> Clarified comments. Please squash commits.

Squashed the commit.

harmandeeps commented:
> > Clarified comments. Please squash commits.
> 
> Squashed the commit.

@losipiuk : The PR build is passed.

losipiuk commented:
Merged thanks
